### PR TITLE
Add Javascript extension and annotation group support

### PR DESCRIPTION
**Pull request by bmedx**
**Description:** Gets the Javascript extension to 90% and adds group / choice support to static annotation checking.

**JIRA:** https://openedx.atlassian.net/browse/PLAT-2348

**Author concerns:** There are still several outstanding things to be done. Much of that work is captured in other groomed tickets. The one big thing the current code doesn't address is multi-line annotations due to some concerns I have with the direction we should take them. Other known issues are marked with TODO's.

### CODE REVIEW

File: code_annotations/extensions/base.py [hunk 1]
```diff
@@ -19,14 +26,219 @@ def __init__(self, config, echo):
...
+            elif isinstance(annotation_tokens[annotation_or_group], (list, tuple)):
+                self._add_annotation_group(annotation_tokens[annotation_or_group])
+            elif isinstance(annotation_tokens[annotation_or_group], dict):
+                self._add_annotation_token(next(iter(annotation_tokens[annotation_or_group])))
+            else:
+                raise TypeError('{} is an unknown type. Annotations must be strings, list/tuples, or dicts.'.format(
+                    annotation_tokens[annotation_or_group])
+                )
+
+    def validate(self, file_handle):  # pragma: no cover
```
Comment (doctoryes) [inline code_annotations/extensions/base.py]: And shouldn't `validate`, `search`, and `_add_annotation_*` all be decorated as `@abstractmethod`?

File: code_annotations/extensions/base.py [hunk 2]
```diff
@@ -19,14 +26,219 @@ def __init__(self, config, echo):
...
+            for match in re.finditer(self.comment_regex, txt):
+                # Should only be one match
+                comment_content = [item for item in match.groups() if item is not None][0]
+                for inner_match in re.finditer(self.query, comment_content):
+                    # No matter how long the regex is, there should only be 2 non-None items,
+                    # with the first being the annotation token and the 2nd being the comment.
+                    cleaned_groups = [item for item in inner_match.groups() if item is not None]
+
+                    if len(cleaned_groups) != 2:  # pragma: no cover
+                        raise Exception('Number of found items in the list is not 2, something is very wrong.')
```
Comment (doctoryes) [inline code_annotations/extensions/base.py]: Nit: On the scale of wrong things, is this `very`? 😺 Maybe adding a print of the cleaned_groups *or* the number of found items would be better?

File: code_annotations/extensions/base.py [hunk 3]
```diff
@@ -19,14 +26,219 @@ def __init__(self, config, echo):
...
+        Raises:
+            TypeError
+        """
+        for annotation in annotation_group:
+            if isinstance(annotation, six.string_types):
+                self._add_annotation_token(annotation)
+            elif isinstance(annotation, dict):
+                annotation_name = next(iter(annotation))
+                self._add_annotation_token(annotation_name)
+            else:
```
Comment (doctoryes) [inline code_annotations/extensions/base.py]: Is there intentionally no `elif` for lists/tuples here? The exception below indicates they should be handled as well.

File: code_annotations/extensions/base.py (hunk 2 above)
Comment (bmedx) [reply code_annotations/extensions/base.py]: I'd classify it as very bad, it means the whole regex is broken and nothing being reported can be trusted, but I also agree that better output is required!

File: code_annotations/extensions/base.py (hunk 3 above)
Comment (bmedx) [reply code_annotations/extensions/base.py]: Opposite problem, that was copypasta. I'll update the exception.

File: code_annotations/extensions/base.py [hunk 4]
```diff
@@ -1,13 +1,20 @@
...
 Abstract and base classes to support plugins.
 """
+import re
+
+import six
+
+from code_annotations.helpers import clean_abs_path
 
 
 class AnnotationExtension(object):
```
[the remark this answers is not in the document]
Comment (bmedx) [reply code_annotations/extensions/base.py]: `ABC` doesn't exist before Python 3.4 so I was avoiding the somewhat confusing metaclass issues by doing it the old school way, but `six` has a way of making it work relatively cleanly so I'll implemented that.

**Pull request merged**